### ISSUE TITLE
A few questions about monitoring services for a research

### ISSUE THREAD

**Issue by SamAI-Software** [raiden-network/raiden-monitoring-service#30]
Hi guys, I'm researching Lightning Network watchtowers (WTs) and Raiden monitoring services (MSs), and writing a series of articles, explaining why they are essential for scaling, how they work, different types of WTs and MSs, their approximate operating and capital costs, scaling challenges, proposed solutions and their trade-offs, viable business models, etc.

If you can answer a few questions, that will be very helpful and I can link this issue in the article, so interested developers can read your answers, join the discussion and potentially contribute to a project.

1. Does monitoring service know which exact channel it watches?
2. There are at least 2 parties (2 on-chain addresses) in the channel. Does monitoring service know which exact party (on-chain address) it watches?
3. Is this info about channel or address included in the each package sent to a monitoring service?
4. Is it possible to implement a logic that a monitoring service will not know exact channel or address it watches with current protocol?
5. If not, what kind of changes to a protocol will be required?
6. How much time user has to broadcast a penalty transaction in case of uncooperative closure? I've seen *2 hours* in some video, if so, then why so low? To increase UX of unilateral closure? But that decreases security.
7. Can the time be easily changed or are there some dependencies that might "break" after that?
8. Can users change this time by mutual agreement upon creation of a channel? Or is it hardcoded on a protocol level?
9. Any approximation on the size (in bytes) of each message (package with latest balance proof) sent to monitoring service after each transaction?
10. Ever growing database of stored messages (packages) can become a bottleneck for scaling. On LN this problem can be solved with eltoo (with privacy trade-offs), which will allow users and watchtowers to store only latest states and thus delete all old states. Is something like this possible on Raiden without any risks of losing funds (e.g. losing multiple pending transfers)?
11. If yes, then does this update/feature has a name? And is it on the roadmap? Any approximate date?
12. Can monitoring services discover the exact values of channel states from received/stored messages? E.g. LN watchtowers don't know the exact state balance, they only know that the state was changed.
13. In case of breaching transaction, do monitoring services broadcast a penalty transaction that moves all channel's funds to a victim's on-chain address, so a cheater can potentially lose some money? Or do monitoring services broadcast a penalty transaction with just latest state, so a cheater doesn't risk anything? 
14. Alice and Bob used a channel for awhile. Alice lost her phone with all balance proofs. She has a recovery seed, so she can access on-chain funds, however, she does not have the latest balance proof. Can Alice get/receive/route transfers using an opened channel with Bob, or should she just sit and wait until he closes the channel unilaterally?
15. I think, I saw something about snapshots and ability to recover balance proofs, but what if there were e.g. 10 more transactions after the latest snapshot?
16. Can Alice (lost her phone with all balance proofs, but has recovery seed) close her channels unilaterally without a risk of losing her funds?
17. Can Alice (lost her phone with all balance proofs, but has recovery seed) retrieve all balance proofs from her monitoring service?
18. Do you have any plans to implement collateral as was described in [PISA](https://twitter.com/paddykcl/status/1071160020404826113) by Patrick McCorry?
19. Any other ideas how to improve accountability of monitoring services?

**Additional:**
This is how LN watchtowers work in a few words.
- After each channel state update, Alice sends to a watchtower encrypted signatures authorizing the movement of all channel's funds back to Alice. 
- The sent package has no info about state's balances.
- Each sent package also contains a "hint" how to find a breaching on-chain transaction.
- Watchtower constantly watches blockchain by checking all new on-chain transactions against its hash table of "hints".
- Once the breaching transaction is found, watchtower is able to decrypt the signatures from Alice's package and reconstruct the penalty transaction that moves all funds back to Alice.

Can you describe how Raiden monitoring services will work in a similar simple way? Or outline the main differences?
**Thanks**!

P.S. I rarely check my github nowadays, so if you read this from the far future, please consider reaching out via [twitter](https://twitter.com/SamAI_Software), [linkedin](https://www.linkedin.com/in/samaisoftware/), or [medium](https://medium.com/@SamAI_Software), if you'll have any questions or info to share.

**Comment by Dominik1999** [comment]
Hi Sam,
 
Thanks for your interest in the Raiden Monitoring Service (MS). We are happy to answer your questions below, however, please note that these answers explain the current envisaged setup for the MS and that the MS is still work in progress. This means that the design might evolve over time. The chosen way forward is intended to be a first way to solve the monitoring problem. This approach is in the research phase and subject to change.

This is how RDN Monitoring Services (MS) will work in a few words **description changed**:
- Whenever Alice wants to get her channel monitored, Alice deposits some tokens to a smart contract and publishes balance proofs including the token network, channel ID, a nonce and hashed balance value into a public chat room
- Updates can be sent whenever the state changes
- The sent balance proof does not expose the actual balance values, as they are hashed
- MSs can pick up these messages and then listen to events regarding this particular channel
- Whenever the channel is closed but not before the settlement period has ended, the MS will call the updateNonClosingBalanceProof function with the provided balance proof. This way the settleChannel function that calculates the token distribution can only be executed submitting the corresponding balance values of the hashed balances provided before 
- Either Alice or Bob (or anyone else) can now anytime call settleChannel and initiate the token distribution
- This will result in a correct distribution of tokens, in accordance with the off-chain transfers
- After a successful submission of the hashed balance by the MS (might be a defeat of an attack), the monitoring service gets an on-chain payment from a smart contract where Alice deposited something beforehand

And about your questions:

1. Does monitoring service know which exact channel it watches? _Yes, it does. The “channel ID” is in the monitoring request. The monitoring service listens to ‘channel close’ events and compares the nonce of the submitted balance proof at closing to the nonce of the balance proof submitted by its customer_
2. There are at least 2 parties (2 on-chain addresses) in the channel. Does the monitoring service know which exact party (on-chain address) it watches? _Yes, it does. It is also the party that pays for the service in the end_
3. Is this info about channel or address included in the each package sent to a monitoring service? _Yes, the customer of the monitoring service wants the MS to know which channel to watch. On every updated balance proof the customer would state the token network and channel ID - with this data, a channel, the involved parties and the initial deposits are fully transparent_
4. Is it possible to implement a logic that a monitoring service will not know exact channel or address it watches with current protocol? _This might be possible, but we don’t plan to implement this at the moment. Happy to discuss your ideas here_
5. If not, what kind of changes to a protocol will be required? _We are happy to discuss this with you in case you have an idea how to do this_
6. How much time a user has to broadcast a penalty transaction in case of uncooperative closure? I've seen 2 hours in some video, if so, then why so low? To increase UX of unilateral closure? But that decreases security. _This is a trade off for every Layer 2 solution in our perspective - we need to have a dispute period in case of misbehavior, but on the other hand the users want to get their tokens out. The duration of the settlement period can be selected by the channel partners individually when opening a channel_
7. Can the time be easily changed or are there some dependencies that might "break" after that? _It can easily be changed_
8. Can users change this time by mutual agreement upon creation of a channel? Or is it hard coded on a protocol level? _It can be set when opening a channel_ 
9. Any approximation on the size (in bytes) of each message (package with latest balance proof) sent to monitoring service after each transaction? _You can find a specification of these messages [here](https://raiden-network-specification.readthedocs.io/en/latest/messaging.html#offchain-balance-proof)_ 
10. Ever growing database of stored messages (packages) can become a bottleneck for scaling. On LN this problem can be solved with eltoo (with privacy trade-offs), which will allow users and watchtowers to store only latest states and thus delete all old states. Is something like this possible on Raiden without any risks of losing funds (e.g. losing multiple pending transfers)? _The monitoring service should delete old states in case it accepts new ones of the same customer and channel_
11. If yes, then does this update/feature has a name? And is it on the roadmap? Any approximate date? _This will be part of the first release_ 
12. Can monitoring services discover the exact values of channel states from received/stored messages? E.g. LN watchtowers don't know the exact state balance, they only know that the state was changed. _No, the exact state will be hashed and the monitoring service can only see that the submitted state of the malicious channel partner does not match the nonce and state it knows_
13. In case of breaching transaction, do monitoring services broadcast a penalty transaction that moves all channel's funds to a victim's on-chain address, so a cheater can potentially lose some money? Or do monitoring services broadcast a penalty transaction with just latest state, so a cheater doesn't risk anything? _No, there is no penalty of losing funds. The Raiden protocol itself is designed in a way that the attacker can only lose money / get less money than he actually would receive when he tries to cheat by submitting an old balance proof (since the transferred amount is a monotonically increasing function)_
14. Alice and Bob used a channel for awhile. Alice lost her phone with all balance proofs. She has a recovery seed, so she can access on-chain funds, however, she does not have the latest balance proof. Can Alice get/receive/route transfers using an opened channel with Bob, or should she just sit and wait until he closes the channel unilaterally? **Changed Answer** _Alice can close without a balance proof. She just will lose money doing so (if she had the BPs, sent a product to Bob and then lost the BPs)_ 
15. I think, I saw something about snapshots and ability to recover balance proofs, but what if there were e.g. 10 more transactions after the latest snapshot? _If the latest balance proofs are lost and the respective channel partner is uncooperative, there is a risk of losing funds_
16. Can Alice (lost her phone with all balance proofs, but has recovery seed) close her channels unilaterally without a risk of losing her funds? _Without having the latest balance proof, there is a risk involved of losing funds_
17. Can Alice (lost her phone with all balance proofs, but has recovery seed) retrieve all balance proofs from her monitoring service? _Theoretically, the monitoring service could provide these. But we currently do not plan this recovery service. However, Alice could not easily settle with these since part of the balance proofs are hashed and can not be used to settle a channel - ~~they can only be used to stop the settlement~~_
18. Do you have any plans to implement collateral as was described in PISA by Patrick McCorry? _We follow all of Patrick’s work and have / had many discussions. At the moment we decided to follow a different approach, but would like to implement PISA in the future_
19. Any other ideas how to improve accountability of monitoring services? _There are other solutions possible, e.g. penalization by losing reputation and therefore customers_

**Comment by SamAI-Software** [comment]
Thanks for the answers, that was **incredibly helpful!**
You can read the first article [**here**](https://medium.com/crypto-punks/lightning-vs-raiden-watchtowers-monitoring-services-differences-c8eb0f724e68).
If you don't mind, we can keep this issue open for awhile, so we can discuss the topic here, and I will probably have more questions for the upcoming articles about watchtowers and monitoring services. If you'll want to participate in a discussion at twitter, here is a [**link**](https://twitter.com/SamAI_Software/status/1080773169605152768).
P.S. Congratulations on the **10th anniversary!** :)

**Comment by Dominik1999** [comment]
@SamAI-Software we discussed the answers here in detail and provided a more exact description. Sorry for the inconvenience. And thank you very much for taking the effort and writing this article

**Comment by SamAI-Software** [comment]
Thanks for letting me know about the changes in the previous description!
I'm writing an article about possible solutions to 3 main issues of watchtowers and monitoring services: privacy, scalability, and accountability.

Current Raiden's monitoring services design follows ["business-oriented" approach](https://medium.com/crypto-punks/lightning-vs-raiden-watchtowers-monitoring-services-differences-c8eb0f724e68), which can scale easier since monitoring services will be able to link together all state updates made by the same user and store only latest state updates, so their data storage requirements will be relatively low comparative to "privacy-oriented" watchtowers. As a result, operating costs are not a big obstacle for scaling, but privacy and accountability are still issues that have to be addressed (ideally, from the start).

**Privacy**
Here is a list of different possible solutions that I'll talk about in the article:
1. **Fuzz transaction time** (e.g., delay sending a state update for a few seconds/minutes). However, in RDN's approach each MS knows which exact channel it watches, so MS knows both parties of a transaction, so both counter-parties have to delay a sending of a state update, otherwise there is not point in it (e.g., Alice has delayed a sending for a few minutes, but Bob has sent immediately). Most laymen won't care about privacy features, so this delay has to be forced on a protocol level. Thoughts?

2. **Rotate watchtowers or monitoring services** after each transaction, so WT/MS won't have a full picture. However, if in RDN user will [send state updates to all registered MSs](https://github.com/raiden-network/raiden-services/issues/20), then there is no value in rotating. Additionally, there most likely be a network of obedient WTs/MSs that will be coerced to participate in a global surveillance. Any ideas?

3. **Bundle together many state updates** from different users in a trustless way similar to coin-join services and then send to MSs in big chunks of data. However, this approach won't increase privacy of current RDN's design, because MSs know which exact channels they're watching, so state updates will still be associated with corresponding on-chain addresses. It might increase privacy a bit though, since MSs will not know the exact time of a state update.

4. **Create lots of network noise** with low-amount fake-transactions (but this approach will not increase privacy in case of a mass-usage of path-finding-services (PFS), since PFSs will be able to determine which transactions were real and which were just a "noise", and then collude with other services in order to participate in a global surveillance. Additionally, laymen won't bother creating fake noise (or even routing), so noise should be forced on a protocol level.

As we can see, these 4 solutions will hardly improve privacy of current RDN's design. Do you have any thoughts about these solutions, or do you have anything else in mind, so we can discuss and I can incorporate into the article?

**Comment by Dominik1999** [comment]
@SamAI-Software thanks for all your effort here! We will come back with some answers / thoughts on your points soon ~ end of this week.

**Comment by Dominik1999** [comment]
Hi @SamAI-Software ! You have anticipated most of the reasoning already, but here are some of our thoughts (thanks to @err508)

- Acc. your point 1. Transaction time could be obfuscated like this, but do you think this is the most critical metadata? It would force users to leave their client running for a random amount of time, when they actually want to shut it down, would you be willing to accept that as a Raiden user?

- Acc. your point 3. In contrast to a coin-join scheme, if a state update bundle as you propose would not be sent in time, this would be problem - so nodes cannot rely on someone else to send the state update to the MS. Additionally, this would again only obfuscate the transaction time, not the content. 

- Acc. your point 4. This proposal is not clear to us: If it is proper noise, how could a monitoring service ever know which state update it is actually supposed to use? 

In summary, we think that real privacy is a hard problem with no easy fix available. Right now, we focus on transaction speed and security rather than privacy, which is a different, but also a hard problem. However, there are many projects working on privacy with number one priority, and we engage with developers and researchers working on ZKSnarks and ZKStarks as privacy is definitely something we want to focus on in the future. 

Thanks again for your thoughts and feedback! Looking forward to your second article

**Comment by palango** [comment]
@SamAI-Software we moved the services to a combined repository, letz's continue the discussion there: https://github.com/raiden-network/raiden-services
